### PR TITLE
Add paths for artificial trees

### PR DESCRIPTION
**Pull request by d367wang**
This is the upstream pull request of https://github.com/opprop/checker-framework-inference/pull/295
This PR combines changes in https://github.com/opprop/checker-framework/pull/133 , and fixes issues introduced by artificial tree paths.

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java [hunk 1]
```diff
@@ -418,6 +419,20 @@ protected void addToWorklist(Block b) {
         }
     }
 
+    public @Nullable UnaryTree getUnaryTreeForAssign(AssignmentNode node) {
```
Comment (d367wang) [inline dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java:422]: 1. add document in dataflow manual & source code
2. consider bi-directional Map
3. IdentityHashMap ?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 2]
```diff
@@ -1719,7 +1720,18 @@ public PhaseOneResult process(CompilationUnitTree root, UnderlyingAST underlying
          *
          * @param tree the newly created Tree
          */
-        public void handleArtificialTree(Tree tree) {}
+        public final void handleArtificialTree(Tree tree) {
```
Comment (d367wang) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java:1723]: Options:
1. Consider lambda
2. move this to CFCFGBuilder, add comment to encourage using this one-argument version

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 3]
```diff
@@ -4839,7 +4851,7 @@ private AssignmentNode createIncrementOrDecrementAssign(
 
             if (target == null) {
                 target = treeBuilder.buildAssignment(exprTree, (ExpressionTree) narrowed.getTree());
-                handleArtificialTree(target);
+                handleArtificialTree(target, TreePath.getPath(getCurrentPath(), exprTree));
```
Comment (d367wang) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java:4854]: add document in dataflow-manual, to demonstrate how this works for postfix operation.

File: framework/src/main/java/org/checkerframework/framework/flow/CFCFGBuilder.java [hunk 4]
```diff
@@ -112,19 +114,23 @@ protected boolean assumeAssertionsEnabledFor(AssertTree tree) {
...
+            MethodTree enclosingMethod = TreeUtils.enclosingMethod(path);
             if (enclosingMethod != null) {
                 Element methodElement = TreeUtils.elementFromDeclaration(enclosingMethod);
                 factory.setEnclosingElementForArtificialTree(tree, methodElement);
             } else {
-                ClassTree enclosingClass = TreeUtils.enclosingClass(getCurrentPath());
+                ClassTree enclosingClass = TreeUtils.enclosingClass(path);
                 if (enclosingClass != null) {
                     Element classElement = TreeUtils.elementFromDeclaration(enclosingClass);
                     factory.setEnclosingElementForArtificialTree(tree, classElement);
```
Comment (d367wang) [inline framework/src/main/java/org/checkerframework/framework/flow/CFCFGBuilder.java:127]: investigate usage of method `setEnclosingElementForArtificialTree`

File: dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java (hunk 1 above)
Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/analysis/AbstractAnalysis.java:422]: 3 is related to 2. There is a https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/HashBiMap.java but I didn't see a IdentityHashBiMap.
And these are related to the `cfg.getUnaryAssignNodeLookup();` call and how to simplify this method.

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java (hunk 3 above)
Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java:4854]: https://github.com/opprop/checker-framework/blob/master/dataflow/manual/content.tex#L979
seems like the right area for some extra explanation.

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 5]
```diff
@@ -3164,13 +3164,20 @@ public final Element getEnclosingElementForArtificialTree(Tree node) {
...
+     *
+     * @param node the {@link Tree} to set the enclosing element to
+     * @param path the {@link TreePath} to attached to {@code node}
+     * @param enclosing the enclosing {@link Element} for {@code node}
      */
-    public final void setEnclosingElementForArtificialTree(Tree node, Element enclosing) {
+    public final void setEnclosingElementForArtificialTree(
+            Tree node, TreePath path, Element enclosing) {
         artificialTreeToEnclosingElementMap.put(node, enclosing);
+        treePathCache.addPath(node, path);
```
Comment (d367wang) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:3180]: new TreePath(path, node)

File: framework/src/main/java/org/checkerframework/framework/flow/CFCFGBuilder.java [hunk 6]
```diff
@@ -114,20 +112,16 @@ protected boolean assumeAssertionsEnabledFor(AssertTree tree) {
...
-            } else {
-                ClassTree enclosingClass = TreeUtils.enclosingClass(getCurrentPath());
-                if (enclosingClass != null) {
-                    Element classElement = TreeUtils.elementFromDeclaration(enclosingClass);
-                    factory.setEnclosingElementForArtificialTree(tree, classElement);
-                }
-            }
+            // Create a new sibling of the current path and assign to the artificial tree
+            TreePath artificialPath = new TreePath(getCurrentPath().getParentPath(), tree);
+            factory.setPathForArtificialTree(tree, artificialPath);
```
Comment (d367wang) [inline framework/src/main/java/org/checkerframework/framework/flow/CFCFGBuilder.java:124]: call handleArtificialTree(Tree tree, TreePath path)

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 7]
```diff
@@ -3153,27 +3148,16 @@ public final TreePath getPath(Tree node) {
...
-     * @param node the {@link Tree} to get the enclosing method for
-     * @return the method {@link Element} enclosing the argument, or null if none has been recorded
-     */
-    public final Element getEnclosingElementForArtificialTree(Tree node) {
-        return artificialTreeToEnclosingElementMap.get(node);
-    }
-
-    /**
-     * Handle an artificial tree by mapping it to the enclosing element.
+     * Set tree path for the given artificial tree.
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:3151]:
```suggestion
     * Set the tree path for the given artificial tree.
```

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 8]
```diff
@@ -1716,6 +1716,15 @@ public PhaseOneResult process(CompilationUnitTree root, UnderlyingAST underlying
...
         }
 
+        /**
+         * Perform any actions required when CFG translation creates a new Tree that is not part of
+         * the original AST, with the corresponding tree path being specified.
+         *
+         * @param tree the newly created Tree
+         * @param path the path for the newly created tree
+         */
+        public void handleArtificialTree(Tree tree, TreePath path) {}
```
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java:1726]: As this method isn't needed any more, please remove it and inline the single call.

File: framework/src/main/java/org/checkerframework/framework/flow/CFCFGBuilder.java [hunk 9]
```diff
@@ -114,20 +112,34 @@ protected boolean assumeAssertionsEnabledFor(AssertTree tree) {
...
-                Element methodElement = TreeUtils.elementFromDeclaration(enclosingMethod);
-                factory.setEnclosingElementForArtificialTree(tree, methodElement);
-            } else {
-                ClassTree enclosingClass = TreeUtils.enclosingClass(getCurrentPath());
-                if (enclosingClass != null) {
-                    Element classElement = TreeUtils.elementFromDeclaration(enclosingClass);
-                    factory.setEnclosingElementForArtificialTree(tree, classElement);
-                }
-            }
+            // Create a new child of the current path and assign to the artificial tree
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/flow/CFCFGBuilder.java:140]: Can you expand this comment a bit and mention the trade-offs between making it a sibling vs child?

File: framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java [hunk 10]
```diff
@@ -364,6 +365,28 @@ public void standardizeVariable(Tree node, AnnotatedTypeMirror type, Element ele
         if (path == null) {
             return;
         }
+
+        if (ele instanceof DetachedVarSymbol) {
+            // Skip standardizing artificial variable trees, because it is the standardized
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java:370]: Can you try to clarify the first sentence?

File: framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java [hunk 11]
```diff
@@ -364,6 +365,28 @@ public void standardizeVariable(Tree node, AnnotatedTypeMirror type, Element ele
...
+            //    static HashMap<Integer, Date> call_hashmap = new HashMap<>();
+            //
+            //    public foo() {
+            //        for (Integer i : call_hashmap.keySet()) {
+            //            @NonNull Date d = call_hashmap.get(i);
+            //        }
+            //    }
+            //
+            // The CFGBuilder creates the initialized artificial iterator as follows:
+            //      iter#num0 = NonNullMapValue.call_hashmap.keySet().iterator()
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java:383]: The `NonNullMapValue` comes as a surprise. I assume that's the class around the above code? Maybe add that there.

File: framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java [hunk 12]
```diff
@@ -364,6 +365,28 @@ public void standardizeVariable(Tree node, AnnotatedTypeMirror type, Element ele
...
+            // Therefore for the map-key type system, the type of the variable "i" is
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java:384]:
```suggestion
            // Therefore, for the Map Key Checker, the type of the variable "i" is
```

**Review by wmdietl — COMMENTED** [review]
A few small comments, but otherwise this is great!

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 13]
```diff
@@ -3806,30 +3782,16 @@ protected final boolean isWithinConstructor(Tree tree) {
...
-    /**
-     * Adds the given mapping from a synthetic (generated) tree to its enclosing element.
+     * Set the tree path for the given artificial tree.
      *
      * <p>See {@code
      * org.checkerframework.framework.flow.CFCFGBuilder.CFCFGTranslationPhaseOne.handleArtificialTree(Tree)}.
      *
-     * @param tree artifical tree
-     * @param enclosing element that encloses {@code tree}
+     * @param node the artificial {@link Tree} to set the path to
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:3790]:
```suggestion
     * @param tree the artificial {@link Tree} to set the path for
```

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 14]
```diff
@@ -3806,30 +3782,16 @@ protected final boolean isWithinConstructor(Tree tree) {
...
+     * @param path the {@link TreePath} for the artificial tree
      */
-    public final void setEnclosingElementForArtificialTree(Tree tree, Element enclosing) {
-        artificialTreeToEnclosingElementMap.put(tree, enclosing);
+    public final void setPathForArtificialTree(Tree node, TreePath path) {
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:3793]:
```suggestion
    public final void setPathForArtificialTree(Tree tree, TreePath path) {
```

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java (hunk 5 above)
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:3794]:
```suggestion
        treePathCache.addPath(tree, path);
```

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 15]
```diff
@@ -2092,9 +2082,9 @@ public AnnotatedNullType getAnnotatedNullType(Set<? extends AnnotationMirror> an
     }
 
     /**
-     * Returns the innermost enclosing method or class tree of {@code tree}. If {@code tree} is
-     * artificial (that is, created by dataflow), then {@link #artificialTreeToEnclosingElementMap}
-     * is used to find the enclosing tree.
+     * Returns the innermost enclosing method or class tree of {@code tree}. Since artificial trees
+     * are assigned to be the child node of the original tree, their enclosing trees are find the
```
[the remark this answers is not in the document]
Comment (wmdietl) [reply framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java:2102]:
```suggestion
     * are assigned to be the child node of the original tree, their enclosing trees are found the
```

**Review by wmdietl — COMMENTED** [review]
Thanks, these look good!
I'm making a few small javadoc tweaks. Please reformat if CI fails.

**Comment by wmdietl** [comment]
Thanks @d367wang !
Can you please open a PR against eisop/checker-framework with this change? I think it is generally useful. Please do the same for your other opprop changes and see which should go into eisop.
And please do remember to close the other PRs that made similar changes

**Pull request merged**